SIMSBIOHUB-470: Observation Table Measurement Validation

al-rosenthal opened the pull request:
## Links to Jira Tickets

[SIMSBIOHUB-470](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-470)

## Description of Changes

- Added front end validation for measurement columns in the observations table
- Refactored validation process and moved files into own validation utils file
- Added measurement validation at API entry level for observation measurements as well
- Fixed bugs in import/ measurement validation process

## Testing Notes
### Setup
- Run Critterbase locally from branch `SIMSBIOHUB-498`
- In SIMS, navigate to the `Manage Observations` table on a survey 
- Add records for Moose (Alces Americanus) and Black capped Chickadee (Poecile atricapillus)
- Save those records 
- Open table Cog menu and add measurements
- Search for antler and add all measurements found
- Search for nest and add all measurements found

### Testing
- Add antler values to the chickadee rows 
- Save changes 
- See validation errors for each cell that has antler data

- Add nest values to Moose rows
- Save Changes
- See validation errors for each cell that has nest data

- Add nest data to the chickadee rows
- Add antler data to the moose rows
- Save changes
- See table refresh with saved data

al-rosenthal committed:
stubbed out validation process for the front end

al-rosenthal committed:
commented out code

al-rosenthal committed:
got basic validation stubbed out

al-rosenthal committed:
returned proper types

al-rosenthal committed:
got some errors wired up to display

al-rosenthal committed:
added proper field keys for error handling

al-rosenthal committed:
validation working with for add and edit

al-rosenthal committed:
validation working

al-rosenthal committed:
added validation utils for observation table

al-rosenthal committed:
swapped out old validation code

al-rosenthal committed:
added documentation

al-rosenthal committed:
wrapped fetch call in a callback

al-rosenthal committed:
added measurement validation to observation insert and update endpoint

al-rosenthal committed:
Merge branch 'dev' into SIMSBIOHUB-470

al-rosenthal committed:
added error dialog if validation fails due to critterbase api

al-rosenthal committed:
fixed broken tests

al-rosenthal committed:
modified measurements into useref

al-rosenthal committed:
fixed issue with validation, renamed variable

al-rosenthal committed:
adding more documentation

al-rosenthal committed:
Merge branch 'dev' into SIMSBIOHUB-470

NickPhura commented:
Messing around with it a bit, seems to work so far. I do need to try it again against some better test data though, was just testing against critterbase dev which doesnt have the best measurements.

There was a todo comment about removing `colDef` from the measurements state. Maybe we can sneak that change in too, if we can in fact remove it everywhere.

al-rosenthal committed:
Merge branch 'dev' into SIMSBIOHUB-470

al-rosenthal committed:
removed coldef from measurement columns type

al-rosenthal committed:
Merge branch 'SIMSBIOHUB-470' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-470

KjartanE committed:
Merge branch 'dev' into SIMSBIOHUB-470

al-rosenthal committed:
removed measurementcolumn type

al-rosenthal committed:
Merge branch 'SIMSBIOHUB-470' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-470

al-rosenthal commented:
> Messing around with it a bit, seems to work so far. I do need to try it again against some better test data though, was just testing against critterbase dev which doesnt have the best measurements.
> 
> There was a todo comment about removing `colDef` from the measurements state. Maybe we can sneak that change in too, if we can in fact remove it everywhere.

I removed the colDef and got things working then took it a step further and just removed the MeasurementColumn type, it was only holding a reference the underlying measurement definition so it seemed silly to keep that as well

NickPhura committed:
Add some more debug logging

NickPhura committed:
Merge branch 'SIMSBIOHUB-470' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-470

al-rosenthal committed:
shifted validation code into observation service

curtisupshall commented on api/src/utils/xlsx-utils/worksheet-utils.ts:
```diff
@@ -499,27 +509,39 @@ export async function getCBMeasurementsFromWorksheet(
         const measurements = await critterBaseService.getTaxonMeasurements(tsn);
         if (!measurements) {
-          // TODO: Any data for a TSN that has no measurements is invalid and should reject the uploaded CSV
+          throw `No measurements found for tsn: ${tsn}`;
```
I would probably only throw an instance of `Error` rather than a string (even though this is technically OK), mostly because a caller that might catch this exception is most likely expecting an instance of `Error`.
```javascript
throw new Error(`No measurements found for tsn: ${tsn}`);
```

NickPhura replied:
+1

curtisupshall commented on api/src/utils/xlsx-utils/worksheet-utils.ts:
```diff
@@ -499,27 +509,39 @@ export async function getCBMeasurementsFromWorksheet(
         tsnMeasurements[tsn] = measurements;
       }
     }
   } catch (error) {
```
Assuming you accept the above comment, you can type anotate this, and refactor the throw on Line 543:
```javascript
} catch (error: Error) {
  throw new ApiGeneralError(`Error connecting to the Critterbase API: ${error.message}`);
}
```

al-rosenthal replied:
the catch error cannot be typed like that, the error value is `any` or `unknown` and there is no guarantee of that the thing thrown is of `Error` type. In a catch like that you'd need do a check like 
`if (error instanceof Error)` to enforce a type on error

curtisupshall replied:
That's fair. Here's a snippet from `KeycloakService` that you might  find inspiration from:
```javascript
    } catch (error) {
      defaultLog.debug({ label: 'getUserByUsername', message: 'error', error: error });
      throw new ApiGeneralError('Failed to get user info from keycloak', [(error as Error).message]);   
    }
```

curtisupshall replied:
Currently it looks like we're just including the entire `error` object in the `ApiGeneralError` constructor, which is fine by me. Feel free to ignore the above comment

NickPhura replied:
Never fully built out the custom error classes, though I think in an ideal world we would always throw some kind of custom known error, just to ensure consistency, and making it more obvious to upstream code what errors could be expected, and what fields are present on the error (as far as labels, messages, error details, etc). I think eventually it would also be nice to have more of a developer error message vs public error message, and maybe even supporting a custom stack concept, so that we can trace an error back to its source even if it gets captured and re-thrown.

curtisupshall commented on app/src/constants/i18n.ts:
```diff
@@ -315,7 +315,12 @@ export const ObservationsTableI18N = {
+  // Fetching TSN Measurements from CritterBase error
+  fetchingTSNMeasurementErrorDialogTitle: 'Error fetching measurement validation',
+  fetchingTSNMeasurementErrorDialogText:
+    'We ran into an error fetching measurement data from Critterbase, taxon selected may not be supported. Please try again. If the error persists, please contact your system administrator.'
```
Might rephrase this to appear less colloquial:
> An error occurred while fetching measurement data from Critterbase. The selected taxon may not be supported. Please try again. If the error persists, please contact your system administrator.

al-rosenthal replied:
updated

curtisupshall commented on app/src/features/surveys/observations/observations-table/observation-row-validation/ObservationRowValidationUtils.ts:
```diff
@@ -0,0 +1,236 @@
+  // Measurement values are invalid, create an error and return
+  return {
+    field,
+    message: `Value provided is outside of the valid range ${minValue} < ${value} < ${maxValue}`
```
Technically the `minValue` and `maxValue` are allowed, so the range is represented by `${minValue} ≤ ${value} ≤ ${maxValue}`

curtisupshall replied:
(Opinion) I probably wouldn't use an inequality to represent the range either way, since if this validation error occurs, the value  of `${value}` wouldn't be within the range, so the inequality wouldn't hold. I would just use [square brackets](https://en.wikipedia.org/wiki/Interval_(mathematics))
```javascript
message: `Value provided is outside of the valid range [${minValue}, ${maxValue}]`
```

al-rosenthal replied:
yeah good point, updated

al-rosenthal committed:
resolved comments on PR

curtisupshall reviewed (approved):
Tested locally, LGTM

NickPhura reviewed (approved):
🐦

al-rosenthal committed:
Merge branch 'dev' into SIMSBIOHUB-470
